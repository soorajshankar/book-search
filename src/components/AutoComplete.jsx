import React, {useEffect, useState} from 'react'

// developer friendly message
// can write logic to remove  `console.debug` in the build script
const defaultMethod = method => e =>
  console.debug(
    `please pass valid ${method} prop function to AutoComplete component to capture this event `,
  )

export default function({
  suggestions = [],
  onChange = defaultMethod('onChange'),
  onSelect = defaultMethod('onSelect'),
  value = '',
  visible = true,
}) {
  const [sValue, setSValue] = useState(value)
  // allows prop change > local state change ie rerender
  useEffect(() => {
    if (sValue !== value) setSValue(value) // value from prop gets priority
  }, [value])

  const onLChange = e => {
    setSValue(e.target.value)
    onChange(e.target.value)
  }
  return (
    <div className="c-ac">
      <input onChange={onLChange} className="c-ac__input" value={sValue} />

      {visible && Array.isArray(suggestions) && suggestions.length > 0 && (
        <div className="c-ac__suggestions">
          {suggestions.map((item, i) => (
            <div
              key={i}
              onClick={() => onSelect(item)}
              className="c-suggestions__item"
            >
              {item}
            </div>
          ))}
        </div>
      )}
    </div>
  )
}
