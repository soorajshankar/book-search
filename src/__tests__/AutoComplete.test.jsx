import React from 'react'
import {render, fireEvent} from '@testing-library/react'
import AutoComplete from '../components/AutoComplete'

describe('renders AutoComplete', () => {
  it('should match Snapshot', () => {
    let count = 0
    const {container} = render(
      <AutoComplete
        value="test"
        suggestions={[{title: 'test'}, {title: 'test2'}]}
        onSelect={() => count++}
      />,
    )
    expect(container).toMatchSnapshot()
  })
  it('should render with values and suggestions', () => {
    let count = 0
    const {container} = render(
      <AutoComplete
        value="test"
        suggestions={[{title: 'test'}, {title: 'test2'}]}
        onSelect={() => count++}
      />,
    )
    expect(container.firstChild.className).toEqual('c-ac')
    expect(container.lastChild.className).toEqual('c-ac')

    // input
    const inputDom = container.querySelector('.c-ac__input')
    expect(inputDom.value).toEqual('test')

    // suggestions
    const listDom = container.querySelector('.c-ac__suggestions')
    expect(listDom.childNodes.length).toEqual(2)
    expect(listDom.firstChild.textContent).toEqual('test')
    expect(listDom.childNodes[1].textContent).toEqual('test2')
  })
  it('should render with no value', () => {
    let count = 0
    const {container} = render(
      <AutoComplete value="" suggestions={[]} onSelect={() => count++} />,
    )
    expect(container.firstChild.className).toEqual('c-ac')
    expect(container.lastChild.className).toEqual('c-ac')
    // input
    const inputDom = container.querySelector('.c-ac__input')
    expect(inputDom.value).toEqual('')

    // suggestions
    const listDom = container.querySelector('.c-ac__suggestions')
    expect(listDom).toBeNull()
  })
  it('should respond to UI events', () => {
    let count = [] //simple hack to replace jest.jn mock
    let value = '' //simple hack to replace jest.jn mock
    const {container} = render(
      <AutoComplete
        value="test"
        suggestions={[{title: 'test'}, {title: 'test2'}]}
        onSelect={d => count.push(d)}
        onChange={e => (value = e)}
      />,
    )
    // suggestions select event
    const button = container.querySelector('.c-suggestions__item')
    fireEvent.click(button.firstChild)
    fireEvent.click(button.firstChild)
    expect(count.length).toEqual(2)

    // input event
    const inputDom = container.querySelector('.c-ac__input')
    fireEvent.change(inputDom, {
      target: {value: 'testdata'},
    })
    // make sure inner component is changed
    expect(inputDom.value).toEqual('testdata')
    // make sure it called onChange prop method (here it will set the variable `value`)
    expect(value).toEqual('testdata')
  })
  it('should rerender with new data and it should be ready for the new events', () => {
    let count = [] //simple hack to replace jest.jn mock
    let value = '' //simple hack to replace jest.jn mock
    const {container, rerender} = render(
      <AutoComplete
        value="test"
        suggestions={[{title: 'test'}, {title: 'test2'}]}
        onSelect={d => count.push(d)}
        onChange={e => (value = e)}
      />,
    )
    // suggestions select event
    const button = container.querySelector('.c-suggestions__item')
    fireEvent.click(button.firstChild)
    fireEvent.click(button.firstChild)
    expect(count.length).toEqual(2)

    // input event
    const inputDom = container.querySelector('.c-ac__input')
    fireEvent.change(inputDom, {
      target: {value: 'testdata'},
    })
    // make sure inner component is changed
    expect(inputDom.value).toEqual('testdata')
    // make sure it called onChange prop method (here it will set the variable `value`)
    expect(value).toEqual('testdata')

    rerender(
      <AutoComplete
        value="rere"
        suggestions={[{title: 'rere'}, {title: 'rere2'}, {title: 'rere3'}]}
        onSelect={d => count.push(d)}
        onChange={e => (value = e)}
      />,
    )

    // input
    // const inputDom = container.querySelector('.c-ac__input')
    expect(inputDom.value).toEqual('rere')

    // suggestions
    const listDom = container.querySelector('.c-ac__suggestions')
    expect(listDom.childNodes.length).toEqual(3)
    expect(listDom.firstChild.textContent).toEqual('rere')
    expect(listDom.childNodes[1].textContent).toEqual('rere2')
    expect(listDom.childNodes[2].textContent).toEqual('rere3')

    // suggestions select event
    fireEvent.click(button.firstChild)
    fireEvent.click(button.firstChild)
    expect(count.length).toEqual(4)

    // input event
    fireEvent.change(inputDom, {
      target: {value: 'testdata2'},
    })
    // make sure inner component is changed
    expect(inputDom.value).toEqual('testdata2')
    // make sure it called onChange prop method (here it will set the variable `value`)
    expect(value).toEqual('testdata2')
  })
})
