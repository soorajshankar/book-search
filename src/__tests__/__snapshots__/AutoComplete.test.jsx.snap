// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AutoComplete should match Snapshot 1`] = `
<div>
  <div
    class="c-ac"
  >
    <input
      class="c-ac__input"
      value="test"
    />
    <div
      class="c-ac__suggestions"
    >
      <div
        class="c-suggestions__item"
      >
        test
      </div>
      <div
        class="c-suggestions__item"
      >
        test2
      </div>
    </div>
  </div>
</div>
`;
